File: today2/surrounded-regions.ts
/**
 * https://leetcode.com/problems/surrounded-regions/
 */
export { solve };

function solve(board: string[][]): void {
  const lastRowIndex = board.length - 1;
  const lastColIndex = board[0].length - 1;

  for (let y = 0; y <= lastColIndex; y++) {
    markIgnore(board, 0, y);
    markIgnore(board, lastRowIndex, y);
  }

  for (let x = 1; x < lastRowIndex; x++) {
    markIgnore(board, x, 0);
    markIgnore(board, x, lastColIndex);
  }

  for (let x = 0; x <= lastRowIndex; x++) {
    for (let y = 0; y <= lastColIndex; y++) {
      if (board[x][y] === 'O') board[x][y] = 'X';
      if (board[x][y] === 'N') board[x][y] = 'O';
    }
  }
}

function markIgnore(board: string[][], x: number, y: number): void {
  if (board[x]?.[y] !== 'O') return;
  board[x][y] = 'N';
  markIgnore(board, x + 1, y);
  markIgnore(board, x - 1, y);
  markIgnore(board, x, y + 1);
  markIgnore(board, x, y - 1);
}

File: python/today10/surrounded_regions.py
# https://leetcode.com/problems/surrounded-regions/
from collections import deque
from typing import List, Tuple

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        rows = len(board)
        cols = len(board[0])

        for col in range(cols):
            self.mark_exclude(board, (0, col))
            self.mark_exclude(board, (rows-1, col))

        for row in range(1, rows-1):
            self.mark_exclude(board, (row, 0))
            self.mark_exclude(board, (row, cols-1))
        
        for row in range(rows):
            for col in range(cols):
                if board[row][col] == 'O':
                    board[row][col] = 'X'
                elif board[row][col] == 'T':
                    board[row][col] = 'O'

    def mark_exclude(self, board: List[List[str]], coord: Tuple[int, int]):
        if board[coord[0]][coord[1]] == 'X':
            return
        
        rows = len(board)
        cols = len(board[0])
        visited = set([coord])
        queue = deque([coord])

        def queue_add(row, col):
            if (row in range(rows) and
                col in range(cols) and
                board[row][col] == 'O' and
                (row,col) not in visited):
                
                visited.add((row, col))
                queue.append((row, col))

        while queue:
            for _ in range(len(queue)):
                row, col = queue.popleft()
                board[row][col]= 'T'
                queue_add(row+1, col)
                queue_add(row-1, col)
                queue_add(row, col+1)
                queue_add(row, col-1)